README.md
# Connector Blockchain

**Connector Blockchain** is a package built on [component](https://github.com/instill-ai/component). It enables [**Instill VDP**](https://github.com/instill-ai/vdp) to connect with blockchain vendors and make your data on chains.

## Highlights
Currently, we support these data vendors:

| Name           | Vendor        | Supported Tasks |
|----------------|---------------|-----------------|
| blockchain-numbers | [NumbersProtocol](https://www.numbersprotocol.io/) | Commit-Asset |

## Contributing

Please refer to the [Contributing Guidelines](./.github/CONTRIBUTING.md) for more details.

## Community support

Please refer to the [community](https://github.com/instill-ai/community) repository.

## License

See the [LICENSE](./LICENSE) file for licensing information.


CHANGELOG.md
# Changelog

## [0.6.0-alpha](https://github.com/instill-ai/connector-blockchain/compare/v0.5.0-alpha...v0.6.0-alpha) (2023-10-13)


### Miscellaneous Chores

* **release:** release v0.6.0-alpha ([4be1e02](https://github.com/instill-ai/connector-blockchain/commit/4be1e02b4ac405c3890ad4d1083c5efd123374fb))

## [0.5.0-alpha](https://github.com/instill-ai/connector-blockchain/compare/v0.4.0-alpha...v0.5.0-alpha) (2023-09-30)


### Miscellaneous Chores

* **release:** release v0.5.0-alpha ([7154061](https://github.com/instill-ai/connector-blockchain/commit/71540610b38e59f283eec0bb3412efc056bad6fb))

## [0.4.0-alpha](https://github.com/instill-ai/connector-blockchain/compare/v0.3.0-alpha...v0.4.0-alpha) (2023-09-13)


### Miscellaneous Chores

* **release:** release v0.4.0-alpha ([5ec5c76](https://github.com/instill-ai/connector-blockchain/commit/5ec5c7691d79acd8a2c58b1a4885eb4f8fc64641))

## [0.3.0-alpha](https://github.com/instill-ai/connector-blockchain/compare/v0.2.0-alpha...v0.3.0-alpha) (2023-08-03)


### Miscellaneous Chores

* **release:** release v0.3.0-alpha ([6badd48](https://github.com/instill-ai/connector-blockchain/commit/6badd48997d1197d86c351afaf2c0e1e31414e93))

## [0.2.0-alpha](https://github.com/instill-ai/connector-blockchain/compare/v0.1.0-alpha...v0.2.0-alpha) (2023-07-20)


### Miscellaneous Chores

* **release:** release v0.2.0-alpha ([0153cef](https://github.com/instill-ai/connector-blockchain/commit/0153cef79d2bffcba7e9e5037a8cb296b943a4ec))

## [0.1.0-alpha](https://github.com/instill-ai/connector-blockchain/compare/v0.1.0-alpha...v0.1.0-alpha) (2023-07-10)


### Features

* add numbers icon config ([#4](https://github.com/instill-ai/connector-blockchain/issues/4)) ([38814a2](https://github.com/instill-ai/connector-blockchain/commit/38814a2536f24de81bb637e64a420937c0a19abb))


### Miscellaneous Chores

* **release:** release v0.1.0-alpha ([d61672c](https://github.com/instill-ai/connector-blockchain/commit/d61672cca5a594b5d2484972b650502f2f150e38))


example/sample/main.go
package main

import (
	"fmt"

	"go.uber.org/zap"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/types/known/structpb"

	"github.com/gofrs/uuid"
	connectorBlockchain "github.com/instill-ai/connector-blockchain/pkg"
)

func main() {

	logger, _ := zap.NewDevelopment()
	// It is singleton, should be loaded when connector-backend started
	connector := connectorBlockchain.Init(logger)

	fmt.Println("dee", connector)
	fmt.Println("Xxx", connector.ListConnectorDefinitions())
	// For apis: Get connector definitsion apis
	for _, v := range connector.ListConnectorDefinitions() {
		b, err := protojson.MarshalOptions{UseProtoNames: true}.Marshal(v)
		fmt.Println(0)
		fmt.Println(string(b), err)
		fmt.Println(1)
		fmt.Println(connector.ListCredentialField(v.Id))
		fmt.Println(2)

	}

	// in connector-backend:
	// if user trigger connectorA
	// ->connectorA.defUid
	// ->connectorA.configuration
	fmt.Println(3)
	execution, _ := connector.CreateExecution(uuid.FromStringOrNil("70d8664a-d512-4517-a5e8-5d4da81756a7"), "default", &structpb.Struct{}, logger)

	fmt.Println(4)
	a, err := execution.ExecuteWithValidation([]*structpb.Struct{})
	fmt.Println(a, err)
	if err != nil {
		fmt.Println(1, err)
	}
	// fmt.Println(4444, connector.GetT())
	// fmt.Println(44441, connector.GetD())

}


pkg/numbers/config/seed/flatten_definitions.py
import jsonref
import json

from os.path import dirname

openapi_schema_template = {
    "openapi": "3.0.0",
    "info": {
        "version": "1.0.0",
        "title": "Numbers Connector Openapi"
    },
    "paths": {
        "/execute": {
            "post": {
                "requestBody": {
                    "required": True,
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "object",
                                "properties": {
                                    "inputs": {
                                        "type": "array",
                                        "items": {
                                            "$ref": "data.json#/task_name/input"
                                        }
                                    }
                                }
                            }
                        }
                    }
                },
                "responses": {
                    "200": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "object",
                                    "properties": {
                                        "outputs": {
                                            "type": "array",
                                            "items": {
                                                "$ref": "data.json#/task_name/output"
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}


def gen_openapi(task_name):
    openapi_schema_encode = json.dumps(openapi_schema_template)
    openapi_schema_encode = openapi_schema_encode.replace(
        "task_name", task_name)
    return json.loads(openapi_schema_encode)


base_path = dirname(__file__)
base_uri = 'file://{}/'.format(base_path)

openapi_schema = {}
openapi_schema['default'] = gen_openapi('default')
with open("./openapi.json", 'w') as openapifile:
    json.dump(openapi_schema, openapifile, indent=2)


with open("./definitions.json") as schema_file:
    a = jsonref.loads(schema_file.read(), base_uri=base_uri,
                      jsonschema=True, merge_props=True)

with open('../definitions.json', 'w') as o:
    json.dump(a, o, indent=2)


pkg/numbers/config/seed/generate_definitions.py
import jsonref
import json
import copy

from os.path import dirname

NAME = "Numbers Protocol"
TASKS = ['default']

DEFINITION_TEMPLATE = [
    {
        "custom": False,
        "documentation_url": "https://www.instill.tech/docs/vdp/blockchain-connectors/numbers",
        "icon": "numbers.svg",
        "icon_url": "",
        "id": "blockchain-numbers",
        "public": True,
        "spec": {},
        "title": NAME,
        "tombstone": False,
        "uid": "70d8664a-d512-4517-a5e8-5d4da81756a7",
        "vendor_attributes": {}
    }
]


OPENAPI_SCHEMA_TEMPLATE = {
    "openapi": "3.0.0",
    "info": {
        "version": "1.0.0",
        "title": f"{NAME} Connector OpenAPI"
    },
    "paths": {
        "/execute": {
            "post": {
                "requestBody": {
                    "required": True,
                    "content": {
                        "application/json": {
                            "schema": {
                                "type": "object",
                                "properties": {
                                    "inputs": {
                                        "type": "array",
                                        "items": {
                                            "$ref": "data.json#/task_name/input"
                                        }
                                    }
                                }
                            }
                        }
                    }
                },
                "responses": {
                    "200": {
                        "description": "",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "type": "object",
                                    "properties": {
                                        "outputs": {
                                            "type": "array",
                                            "items": {
                                                "$ref": "data.json#/task_name/output"
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}


def gen_openapi(task_name):
    openapi_schema_encode = json.dumps(OPENAPI_SCHEMA_TEMPLATE)
    openapi_schema_encode = openapi_schema_encode.replace(
        "task_name", task_name)
    return json.loads(openapi_schema_encode)


def convert_field(field):
    field = copy.deepcopy(field)
    if "const" in field:
        return field

    if "type" not in field or field["type"] == "object":
        if "properties" in field:
            for k, v in field["properties"].items():
                field["properties"][k] = convert_field(v)
        if "allOf" in field:
            for i in range(len(field["allOf"])):
                field["allOf"][i] = convert_field(field["allOf"][i])
        if "anyOf" in field:
            for i in range(len(field["anyOf"])):
                field["anyOf"][i] = convert_field(field["anyOf"][i])
        if "oneOf" in field:
            for i in range(len(field["oneOf"])):
                field["oneOf"][i] = convert_field(field["oneOf"][i])
    else:
        if "instillFormat" not in field:
            return field
        assert ("instillFormat" in field)
        assert ("instillUpstreamTypes" in field)
        instill_upstream_types = field["instillUpstreamTypes"]
        original_field = copy.deepcopy(field)
        original_field.pop("title", None)
        original_field.pop("description", None)
        original_field.pop("instillFormat", None)
        original_field.pop("instillUpstreamTypes", None)

        field = {
            "title": field["title"],
            "description": field.get("description", ""),
            "instillFormat": field["instillFormat"],
            "instillUpstreamTypes": field["instillUpstreamTypes"],
            "anyOf": []
        }

        for instillUpstreamType in instill_upstream_types:
            if instillUpstreamType == "value":
                original_field["instillUpstreamType"] = instillUpstreamType
                field["anyOf"].append(original_field)
            if instillUpstreamType == "reference":
                field["anyOf"].append({
                    "type": "string",
                    "pattern": "^\\{.*\\}$",
                    "instillUpstreamType": instillUpstreamType
                })

    return field


def convert_data_schema_to_component_schema(data):
    component_schema = {
        "$schema": "http://json-schema.org/draft-07/schema#",
        "title": f"{NAME} Component Setting",
        "type": "object"
    }
    if len(TASKS) > 1:
        component_schema["properties"] = {
            "input": {
                "properties": {
                    "task": {
                        "title": "Task",
                        "enum": TASKS
                    }
                }
            }
        }
        component_schema["allOf"] = []
        for task in TASKS:
            data[task]["input"]["title"] = "Data Flow"

            component_schema["allOf"].append({
                "if": {
                    "properties": {
                        "input": {
                            "properties": {
                                "task": {
                                    "const": task
                                }
                            }
                        }
                    }
                },
                "then": {
                    "properties": {
                        "metadata": {
                            "title": "Metadata",
                            "type": "object"
                        },
                        "input": convert_field(data[task]["input"])
                    }
                }
            })
    else:
        data["default"]["input"]["title"] = "Data Flow"
        component_schema["properties"] = {
            "metadata": {
                "title": "Metadata",
                "type": "object"
            },
            "input": convert_field(data["default"]["input"])
        }


    return component_schema


if __name__ == "__main__":

    base_path = dirname(__file__)
    base_uri = "file://{}/".format(base_path)

    openapi_schema = {}
    for task in TASKS:
        openapi_schema[task] = gen_openapi(task)

    with open("./resource.json", "r") as rsc_file:
        resource = json.load(rsc_file)
    with open("./data.json", "r") as data_file:
        data = jsonref.load(data_file, base_uri=base_uri,
                            jsonschema=True, merge_props=True)

    DEFINITION_TEMPLATE[0]["spec"]["resource_specification"] = resource
    DEFINITION_TEMPLATE[0]["spec"]["component_specification"] = convert_data_schema_to_component_schema(
        data)
    DEFINITION_TEMPLATE[0]["spec"]["openapi_specifications"] = openapi_schema

    a = jsonref.loads(json.dumps(DEFINITION_TEMPLATE), base_uri=base_uri,
                      jsonschema=True, merge_props=True)

    with open("../definitions.json", "w") as o:
        out = json.dumps(a, indent=2)
        o.write(out)
        o.write("\n")
